File: src/examples/sql_restart/sql_restart.cpp
// Solve the problem 
//
// min x + y + z
// st  [ x y; y z] >= 0 <==> xz >= y^2
//     y >=Q z <==> y >= |z|
//     z >= 1
//
// which should have an optimal solution of (1,1,1).  Basically, we have
// a problem with a semidefinite, quadratic, and linear cone.  Then, we're
// going to run a few iterations, stop the optimization, write the restart,
// load the restart, and then continue with the optimization.  This is basically
// a way to both show a bunch of features and add another unit test.

#include <iostream>
#include <iomanip>
#include "optizelle/optizelle.h"
#include "optizelle/vspaces.h"
#include "optizelle/json.h"

// Define a simple objective where 
//  
// f(x,y,z)=x+y+z
//
struct MyObj : public Optizelle::ScalarValuedFunction <double,Optizelle::Rm> {
    typedef double Real;
    typedef Optizelle::Rm <Real> X;
    
    // Evaluation 
    double eval(X::Vector const & x) const {
        auto sum = Real(0.);
        for(auto const & ele: x) 
            sum += ele;
        return sum;
    }

    // Gradient
    void grad(
        X::Vector const & x,
        X::Vector & grad 
    ) const {
        for(auto & ele : grad)
            ele = Real(1.);
    }

    // Hessian-vector product
    void hessvec(
        X::Vector const & x,
        X::Vector const & dx,
        X::Vector & H_dx
    ) const {
        X::zero(H_dx);
    }
};

// Define a simple SQL inequality 
//
// h(x,y,z) = [ x y ] >=S 0
//            [ y z ]
//            y >=Q z
//            z >=L 1
//
struct MyIneq 
    : public Optizelle::VectorValuedFunction
        <double,Optizelle::Rm,Optizelle::SQL> 
{
    typedef Optizelle::Rm <double> X;
    typedef Optizelle::SQL <double> Z;
    typedef double Real;

    // z=h(x) 
    void eval(
        X::Vector const & x,
        Z::Vector & z
    ) const {
        z(1,1,1)=x[0];
        z(1,1,2)=x[1];
        z(1,2,1)=x[1];
        z(1,2,2)=x[2];

        z(2,1)=x[1];
        z(2,2)=x[2];

        z(3,1)=x[2]-1;
    }

    // y=h'(x)dx
    void p(
        X::Vector const & x,
        X::Vector const & dx,
        Z::Vector & z
    ) const {
        z(1,1,1)=dx[0];
        z(1,1,2)=dx[1];
        z(1,2,1)=dx[1];
        z(1,2,2)=dx[2];

        z(2,1)=dx[1];
        z(2,2)=dx[2];

        z(3,1)=dx[2];
    }

    // x_hat=h'(x)*dz
    void ps(
        X::Vector const & x,
        Y::Vector const & dz,
        X::Vector & x_hat 
    ) const {
        X::zero(x_hat);

        // Remember, the input to this function may not be symmetric, so
        // compute accordingly 
        x_hat[0] += dz(1,1,1);
        x_hat[1] += dz(1,1,2)+dz(1,2,1);
        x_hat[2] += dz(1,2,2);

        x_hat[1] += dz(2,1);
        x_hat[2] += dz(2,2);

        x_hat[2] += dz(3,1);
    }

    // x_hat=(h''(x)dx)*dy
    void pps(
        X::Vector const & x,
        X::Vector const & dx,
        Y::Vector const & dz,
        X::Vector & x_hat 
    ) const {
        X::zero(x_hat);
    }
};

int main(int argc,char* argv[]){
    // Create some type shortcuts
    typedef Optizelle::Rm <double> X;
    typedef Optizelle::SQL <double> Z;
    typedef X::Vector X_Vector;
    typedef Z::Vector Z_Vector;

    // Read in the name for the input file
    if(argc!=2) {
        std::cerr << "sql_restart <parameters>" << std::endl;
        exit(EXIT_FAILURE);
    }
    auto fname = argv[1];

    // Generate an initial guess for the primal
    auto x = std::vector <double> { 5.5, 3.3, 2.2};
    
    // Generate an initial guess for the dual
    auto z = Z_Vector(
        {Optizelle::Cone::Semidefinite,
         Optizelle::Cone::Quadratic,
         Optizelle::Cone::Linear},
        {2,2,1});

    // Create an optimization state
    Optizelle::InequalityConstrained <double,Optizelle::Rm,Optizelle::SQL>
        ::State::t state(x,z);
    
    // Read the parameters from file
    Optizelle::json::InequalityConstrained <double,Optizelle::Rm,Optizelle::SQL>
        ::read(fname,state);

    // Create a bundle of functions
    Optizelle::InequalityConstrained <double,Optizelle::Rm,Optizelle::SQL>
        ::Functions::t fns;
    fns.f.reset(new MyObj);
    fns.h.reset(new MyIneq);

    // Solve the optimization problem
    Optizelle::InequalityConstrained <double,Optizelle::Rm,Optizelle::SQL>
        ::Algorithms::getMin(Optizelle::Messaging::stdout,fns,state);
    
    // Write an intermediate restart file 
    Optizelle::json::InequalityConstrained <double,Optizelle::Rm,Optizelle::SQL>
        ::write_restart("restart.json",state);

    // Read in the restart file
    Optizelle::json::InequalityConstrained <double,Optizelle::Rm,Optizelle::SQL>
        ::read_restart("restart.json",x,z,state);

    // Change the maximum number of iterations to something larger and reset
    // the convergence flag
    state.iter_max = 500;
    state.opt_stop = Optizelle::OptimizationStop::NotConverged;

    // Finish solving the optimization problem
    Optizelle::InequalityConstrained <double,Optizelle::Rm,Optizelle::SQL>
        ::Algorithms::getMin(Optizelle::Messaging::stdout,fns,state);

    // Print out the reason for convergence
    std::cout << "The algorithm converged due to: " <<
        Optizelle::OptimizationStop::to_string(state.opt_stop) << std::endl;

    // Print out the final answer
    std::cout << std::setprecision(16) << std::scientific 
        << "The optimal point is: (" << state.x[0] << ','
        << state.x[1] << ',' << state.x[2] << ')' << std::endl;

    // Write out the final answer to file
    Optizelle::json::InequalityConstrained <double,Optizelle::Rm,Optizelle::SQL>
        ::write_restart("solution.json",state);

    // Successful termination
    return EXIT_SUCCESS;
}

File: src/examples/sql_restart/sql_restart.m
% Solve the problem %
% min x + y + z
% st  [ x y; y z] >= 0 <==> xz >= y^2
%     y >=Q z <==> y >= |z|
%     z >= 1
%
% which should have an optimal solution of (1,1,1).  Basically, we have
% a problem with a semidefinite, quadratic, and linear cone.  Then, we're
% going to run a few iterations, stop the optimization, write the restart,
% load the restart, and then continue with the optimization.  This is basically
% a way to both show a bunch of features and add another unit test.

function sql_restart(fname)
    % Read in the name for the input file
    if nargin ~=1
        error('sql_restart <parameters>');
    end

    % Execute the optimization
    main(fname);
end

% Define a simple objective where 
% 
% f(x,y,z)=x+y+z
%
function self = MyObj()

    % Evaluation 
    self.eval = @(x) sum(x);

    % Gradient
    self.grad = @(x) ones(size(x)); 

    % Hessian-vector product
    self.hessvec = @(x,dx) zeros(size(x)); 
end

% Define a simple SQL inequality 
%
% h(x,y,z) = [ x y ] >=S 0
%            [ y z ]
%            y >=Q z
%            z >=L 1
%
function self = MyIneq()

    % z=h(x) 
    self.eval = @(x)MyIneq_eval(x);

    % z=h'(x)dx
    self.p = @(x,dx)MyIneq_p(x,dx);

    % xhat=h'(x)*dz
    self.ps = @(x,dz)MyIneq_ps(x,dz);

    % xhat=(h''(x)dx)*dz
    self.pps = @(x,dx,dz) zeros(size(x)); 
end

% z=h(x) 
function z=MyIneq_eval(x)
    global Optizelle;
    z = Optizelle.SQL.create( ...
        [Optizelle.Cone.Semidefinite, ...
         Optizelle.Cone.Quadratic, ...
         Optizelle.Cone.Linear], ...
        [2,2,1]);
    z.data{1} = [
        x(1) x(2);
        x(2) x(3)];
    z.data{2} = [x(2);x(3)];
    z.data{3} = x(3)-1;
end

% z=h'(x)dx
function z=MyIneq_p(x,dx)
    global Optizelle;
    z = Optizelle.SQL.create( ...
        [Optizelle.Cone.Semidefinite, ...
         Optizelle.Cone.Quadratic, ...
         Optizelle.Cone.Linear], ...
        [2,2,1]);
    z.data{1} = [
        dx(1) dx(2);
        dx(2) dx(3)];
    z.data{2} = [dx(2);dx(3)];
    z.data{3} = dx(3);
end

% x_hat=h'(x)*dz
function x_hat=MyIneq_ps(x,dz)
    % Remember, the input to this function may not be symmetric, so
    % compute accordingly 
    x_hat = ...
        [dz.data{1}(1,1); ...
         dz.data{1}(1,2) + dz.data{1}(2,1); ...
         dz.data{1}(2,2)] + ...
        [0;dz.data{2}] + ...
        [0;0;dz.data{3}];
end

% Actually runs the program
function main(fname)

    % Grab the Optizelle library
    global Optizelle;
    setupOptizelle();

    % Generate an initial guess for the primal
    x = [5.5; 3.3; 2.2];

    % Generate an initial guess for the dual
    %---SQLSpec0---
    types = ...
        [Optizelle.Cone.Semidefinite, ...
         Optizelle.Cone.Quadratic, ...
         Optizelle.Cone.Linear];
    sizes = [2,2,1];
    %---SQLSpec1---
    %---SQLVector0---
    z = Optizelle.SQL.create(types,sizes);
    %---SQLVector1---

    % Create an optimization state
    state=Optizelle.InequalityConstrained.State.t( ...
        Optizelle.Rm,Optizelle.SQL,x,z);

    % Read the parameters from file
    state=Optizelle.json.InequalityConstrained.read( ...
        Optizelle.Rm,Optizelle.SQL,fname,state);
    
    % Create a bundle of functions
    fns=Optizelle.InequalityConstrained.Functions.t;
    fns.f=MyObj();
    fns.h=MyIneq();

    % Solve the optimization problem
    state=Optizelle.InequalityConstrained.Algorithms.getMin( ...
        Optizelle.Rm,Optizelle.SQL,Optizelle.Messaging.stdout,fns,state);
    
    % Write an intermediate restart file 
    Optizelle.json.InequalityConstrained.write_restart( ...
        Optizelle.Rm,Optizelle.SQL,'restart.json',state);

    % Read in the restart file
    state = Optizelle.json.InequalityConstrained.read_restart( ...
        Optizelle.Rm,Optizelle.SQL,'restart.json',state.x,state.z);

    % Change the maximum number of iterations to something larger and reset
    % the convergence flag
    state.iter_max = 500;
    state.opt_stop = Optizelle.OptimizationStop.NotConverged;

    % Finish solving the optimization problem
    state=Optizelle.InequalityConstrained.Algorithms.getMin( ...
        Optizelle.Rm,Optizelle.SQL,Optizelle.Messaging.stdout,fns,state);

    % Print out the reason for convergence
    fprintf('The algorithm converged due to: %s\n', ...
        Optizelle.OptimizationStop.to_string(state.opt_stop));

    % Print out the final answer
    fprintf('The optimal point is: (%e,%e,%e)\n', ...
        state.x(1),state.x(2),state.x(3));

    % Write out the final answer to file
    Optizelle.json.InequalityConstrained.write_restart( ...
        Optizelle.Rm,Optizelle.SQL,'solution.json',state);
end
